### PR TITLE
[Dataset] Citation graph

### PR DESCRIPTION
**Pull request by classicsong**
## Description
Citation dataset with new interface.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

File: python/dgl/data/citation_graph.py [hunk 1]
```diff
@@ -11,10 +11,17 @@
...
+from .utils import generate_mask_tensor
 from ..utils import retry_method_with_fix
+from .. import backend as F
 from ..graph import DGLGraph
 from ..graph import batch as graph_batch
+from ..convert import to_networkx
+
+backend = os.environ.get('DGLBACKEND', 'pytorch')
 
 _urls = {
```
Comment (VoVAllen) [inline python/dgl/data/citation_graph.py:26]: can we make _urls as a class attribute inside CitationGraphDataset?

File: python/dgl/data/citation_graph.py [hunk 2]
```diff
@@ -166,160 +464,100 @@ def _sample_mask(idx, l):
     mask[idx] = 1
     return mask
 
-def load_cora():
-    data = CitationGraphDataset('cora')
-    return data
+"""Get CoraGraphDataset
```
Comment (VoVAllen) [inline python/dgl/data/citation_graph.py:467]: Function comment should be inside function
def load_cora(...):
    """Get....."""

File: python/dgl/data/utils.py [hunk 3]
```diff
@@ -14,6 +14,10 @@
 from .graph_serialize import save_graphs, load_graphs, load_labels
 from .tensor_serialize import save_tensors, load_tensors
 
+from .. import backend as F
+
+backend = os.environ.get('DGLBACKEND', 'pytorch')
```
Comment (VoVAllen) [inline python/dgl/data/utils.py:19]: Use dgl.backend.backend_name instead

File: python/dgl/data/citation_graph.py [hunk 4]
```diff
@@ -29,34 +36,38 @@ def _pickle_load(pkl_file):
...
+        Raw file directory to download/contains the input data directory.
+        Default: ~/.dgl/
+    force_reload : bool
+        Whether to reload the dataset. Default: False
+    verbose: bool
+      Whether to print out progress information. Default: True.
     """
-    def __init__(self, name):
+    def __init__(self, name, raw_dir=None, force_reload=False, verbose=True):
         assert name.lower() in ['cora', 'citeseer', 'pubmed']
```
Comment (HuXiangkun) [inline python/dgl/data/citation_graph.py:56]: Can CoauthorDataset and AmazonCoBuyDataset in gnn_benchmark dataset use 'name' to identify different part of the dataset? 
For CoauthorDataset, there are two domains 'cs' and 'physics', now I split the dataset as CoauthorCSDataset and CoauthorPhysicsDataset.

File: python/dgl/data/citation_graph.py [hunk 5]
```diff
@@ -113,37 +124,324 @@ def _load(self):
...
+    """ Citation graph is used in many examples
+        We preserve these properties for compatability.
+    """
+    @property
+    def graph(self):
+        return self._graph
+
+    @property
+    def train_mask(self):
+        return self.g.ndata['train_mask']
```
Comment (HuXiangkun) [inline python/dgl/data/citation_graph.py:231]: Do we need add deprecate warning here?

File: python/dgl/data/citation_graph.py [hunk 6]
```diff
@@ -113,37 +124,324 @@ def _load(self):
...
+    
+    Nodes mean paper and edges mean citation 
+    relationships. Each node has a predefined 
+    feature with 1433 dimensions. The dataset is 
+    designed for the node classification task. 
+    The task is to predict the category of 
+    certain paper.
+
+    Statistics
+    ===
```
Comment (HuXiangkun) [inline python/dgl/data/citation_graph.py:260]: Instead of '===' we need to use '-----'.

File: python/dgl/data/citation_graph.py [hunk 7]
```diff
@@ -113,37 +124,324 @@ def _load(self):
...
+    def save(self):
+        """save the graph list and the labels"""
+        graph_path = os.path.join(self.save_path,
+                                  self.save_name + '.bin')
+        info_path = os.path.join(self.save_path,
+                                 self.save_name + '.pkl')
+        save_graphs(str(graph_path), self.g)
+        save_info(str(info_path), {'num_labels': self.num_labels})
+        if self.verbose:
+            print('Done saving data into cached files.')
```
Comment (HuXiangkun) [inline python/dgl/data/citation_graph.py:178]: This is output info has been implemented in the DGLDataset _load().

File: python/dgl/data/dgl_dataset.py [hunk 8]
```diff
@@ -182,15 +180,21 @@ class DGLBuiltinDataset(DGLDataset):
...
-    def __init__(self, name, url, force_reload=False):
+    def __init__(self, name, url, raw_dir=None, force_reload=False, verbose=False):
         super(DGLBuiltinDataset, self).__init__(name,
                                                 url=url,
-                                                raw_dir=None,
+                                                raw_dir=raw_dir,
                                                 save_dir=None,
-                                                force_reload=force_reload)
+                                                force_reload=force_reload,
+                                                verbose=False)
```
Comment (HuXiangkun) [inline python/dgl/data/dgl_dataset.py:197]:
```suggestion
                                                verbose=verbose)
```

File: python/dgl/data/citation_graph.py (hunk 1 above)
Comment (classicsong) [reply python/dgl/data/citation_graph.py:26]: yes

File: python/dgl/data/synthetic_graph.py [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+        mask_array = np.zeros((num_nodes,), dtype=np.int32)
+        mask_array[0:ntrain] = 1
+        mask_array[ntrain:ntrain+nval] = 2
+        mask_array[ntrain+nval:ntrain+nval+ntest] = 3
+        rng.shuffle(mask_array)
+        self.train_mask = (mask_array == 1).astype(np.int32)
+        self.val_mask = (mask_array == 2).astype(np.int32)
+        self.test_mask = (mask_array == 3).astype(np.int32)
+
+        print('Finished synthetic dataset generation.')
```
Comment (hetong007) [inline python/dgl/data/synthetic_graph.py:45]: Need to check `self.verbose` for all `print`

File: python/dgl/data/citation_graph.py [hunk 10]
```diff
@@ -166,160 +466,101 @@ def _sample_mask(idx, l):
...
+    -----------
+        raw_dir : str
+            Raw file directory to download/contains the input data directory.
+            Default: ~/.dgl/
+        force_reload : bool
+            Whether to reload the dataset. Default: False
+        verbose: bool
+        Whether to print out progress information. Default: True.
+        
+        Returns
```
Comment (HuXiangkun) [inline python/dgl/data/citation_graph.py:483]:
```suggestion
    Returns
```

File: python/dgl/data/citation_graph.py [hunk 11]
```diff
@@ -166,160 +466,101 @@ def _sample_mask(idx, l):
...
+        ===
```
Comment (HuXiangkun) [inline python/dgl/data/citation_graph.py:484]:
```suggestion
    -------
```

File: python/dgl/data/citation_graph.py [hunk 12]
```diff
@@ -166,160 +466,101 @@ def _sample_mask(idx, l):
...
+    -----------
+        raw_dir : str
+            Raw file directory to download/contains the input data directory.
+            Default: ~/.dgl/
+        force_reload : bool
+            Whether to reload the dataset. Default: False
+        verbose: bool
+        Whether to print out progress information. Default: True.
+        
+        Returns
```
Comment (HuXiangkun) [inline python/dgl/data/citation_graph.py:503]:
```suggestion
    Returns
```

File: python/dgl/data/citation_graph.py [hunk 13]
```diff
@@ -166,160 +466,101 @@ def _sample_mask(idx, l):
...
+        ===
```
Comment (HuXiangkun) [inline python/dgl/data/citation_graph.py:504]:
```suggestion
    -------
```

File: python/dgl/data/citation_graph.py [hunk 14]
```diff
@@ -166,160 +466,101 @@ def _sample_mask(idx, l):
...
+    -----------
+        raw_dir : str
+            Raw file directory to download/contains the input data directory.
+            Default: ~/.dgl/
+        force_reload : bool
+            Whether to reload the dataset. Default: False
+        verbose: bool
+        Whether to print out progress information. Default: True.
+        
+        Returns
```
Comment (HuXiangkun) [inline python/dgl/data/citation_graph.py:523]:
```suggestion
    Returns
```

File: python/dgl/data/citation_graph.py [hunk 15]
```diff
@@ -166,160 +466,101 @@ def _sample_mask(idx, l):
...
+        ===
```
Comment (HuXiangkun) [inline python/dgl/data/citation_graph.py:524]:
```suggestion
    -------
```

**Comment by HuXiangkun** [comment]
LGTM

**Pull request merged**